File: packages/shared/src/types/presence.ts
import { z } from 'zod'

export const PresenceStatus = z.enum(['online', 'offline', 'busy'])
export type PresenceStatus = z.infer<typeof PresenceStatus>

export const Presence = z.object({
  handle: z.string(),
  status: PresenceStatus,
  custom_message: z.string().nullable(),
  last_seen: z.string().datetime().nullable(),
})
export type Presence = z.infer<typeof Presence>

export const PresenceUpdate = z.object({
  status: PresenceStatus,
  custom_message: z.string().max(200).optional(),
})
export type PresenceUpdate = z.infer<typeof PresenceUpdate>

/** POST /v1/presence/batch — query up to 100 handles at once */
export const PresenceBatchRequest = z.object({
  handles: z.array(z.string()).min(1).max(100),
})
export type PresenceBatchRequest = z.infer<typeof PresenceBatchRequest>

/** Wire shape pushed over WS on presence.update events */
export const PresenceBroadcast = z.object({
  handle: z.string(),
  status: PresenceStatus,
  custom_message: z.string().nullable(),
})
export type PresenceBroadcast = z.infer<typeof PresenceBroadcast>

File: packages/sdk-python/src/agentchat/types/presence.py
"""Presence wire types."""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict

PresenceStatus = Literal["online", "offline", "busy"]


class _BaseModel(BaseModel):
    model_config = ConfigDict(extra="allow", populate_by_name=True)


class Presence(_BaseModel):
    handle: str
    status: PresenceStatus
    custom_message: Optional[str] = None
    last_seen: Optional[str] = None


class PresenceUpdate(_BaseModel):
    status: PresenceStatus
    custom_message: Optional[str] = None


class PresenceBatchRequest(_BaseModel):
    """POST /v1/presence/batch — up to 100 handles at once."""

    handles: List[str]


class PresenceBroadcast(_BaseModel):
    """Wire shape pushed over WebSocket on ``presence.update`` events."""

    handle: str
    status: PresenceStatus
    custom_message: Optional[str] = None
